Sakuli/sakuli.GitHub.io#127/create migration page (#49)
sakuli/docs: content/writing_tests/migration-sakuli-1.md

---
title : "Migration from Sakuli 1"
date :  2020-05-20T08:56:58+02:00

> weight : 10

Reviewer: What's the reason for `weight: 10`?

I also noticed the ordering of posts is off by one and starts at `2`

Author: Just to ensure it stays at the bottom :smile: 
AC in https://github.com/sakuli/sakuli.github.io/issues/127

Author: Altered the weight to 8

---

# How to Migrate from Sakuli 1

With Sakuli v2+ we changed a lot under the hood and completely replaced the technology-stack. This resulted in a more flexible and future-ready architecture while preserving a maximum of backward compatibility. Nearly all functions and classes from Sakuli v1 are available in Sakuli v2+. They are provided by the [@sakuli/legacy](/apidoc/sakuli-legacy/interfaces/legacyapi) package.

### Async / Await

You might have wondered where the `await` keyword came from when you saw the Sakuli v2+ code. This resulted from the decision to use Node.js as the new runtime for Sakuli. The following section will describe the background of the async/await syntax and its use in Sakuli. If you are already familiar with JavaScript, you can skip this section since it contains a lot of basic information.

#### Why we need it

***TL; DR;***

- Due to the runtime and core libraries, asynchronous operations are required (and cannot be turned into synchronous operations)
- Async/Await is the most idiomatic way to handle asynchronous code in JavaScript

> If you ever worked with Sakuli and followed our ["Getting started"](/docs/getting-started/) guide, you might have noticed one of the most obvious paradigmatic changes regarding the test syntax: The wrapping `(async () => /*...*/)();` and the extensive use of `await`. This is due to the new runtime <a href="https://nodejs.org" target="_blank">Node.js</a> and its asynchronous nature. Node.js executes scripts in a single-thread and therefore makes heavy use of asynchronous operations avoiding to block the execution within the thread. The underlying JavaScript engine is <a href="https://v8.dev/" target="_blank">V8</a> (which also powers the JavaScript execution in <a href="https://www.chromium.org/Home" target="_blank">Chromium Browsers</a>). To make this asynchronous behavior possible, V8 uses a so called "Event-Loop" with a "Call-Stack" containing all deferred operations (like: reading a file, making a web-request, doing heavy computation and so on). <a href="https://www.youtube.com/watch?v=8aGhZQkoFbQ" target="_blank">This talk</a> explains this behavior in a short and entertaining way. A good illustration for the event loop is a (fast-food) restaurant, where you place an order at the counter and get a receiver that informs you when your food is ready. Now you can do something else in the meantime while not blocking other processes or being blocked by them until you get your food and can do further operations (e.g. eat your meal).

Reviewer: ```suggestion
If you ever worked with Sakuli and followed our ["Getting started"](/docs/getting-started/) guide, you might have noticed one of the most obvious paradigm changes regarding the test syntax: The wrapping `(async () => /*...*/)();` and the extensive use of `await`. This is due to the new runtime <a href="https://nodejs.org" target="_blank">Node.js</a> and its asynchronous nature. Node.js executes scripts in a single thread and therefore makes heavy use of asynchronous operations, avoiding to block the execution on this thread. The underlying JavaScript engine is <a href="https://v8.dev/" target="_blank">V8</a> (which also powers the JavaScript execution in <a href="https://www.chromium.org/Home" target="_blank">Chromium Browsers</a>). To make this asynchronous behavior possible, V8 uses a so called "Event-Loop" with a "Call-Stack" containing all deferred operations (like: reading a file, making a web-request, doing heavy computation and so on). <a href="https://www.youtube.com/watch?v=8aGhZQkoFbQ" target="_blank">This talk</a> explains this behavior in a short and entertaining way. A good illustration for the event loop is a (fast-food) restaurant, where you place an order at the counter and get a receiver that informs you when your food is ready. Now you can do something else in the meantime while not blocking other processes or being blocked by them until you get your food and can do further operations (e.g. eat your meal).
```

> But how does all this effect Sakuli? Sakuli v1 was running within a Java Virtual Machine and the test scripts were executed within the <a href="https://developer.mozilla.org/de/docs/Rhino" target="_blank">V8</a> (JavaScript Runtime for the JVM) - its fundamental treatment of asynchronous operations completely differs compared to Node. Simply said, everything in Java is blocking until you make it non-blocking, while in Node (many) operations are non-blocking by default. In Java for example it is a rather common custom to run while-loops until the asynchronous operation sets its condition to false. This repl displays the basic idea: <a href="https://repl.it/@tnobody/non-blocking-vs-blockingjava" target="_blank">V8</a>.

Reviewer: ```suggestion
But how does all this effect Sakuli? Sakuli v1 was running within a Java Virtual Machine and the test scripts were executed within the <a href="https://developer.mozilla.org/de/docs/Rhino" target="_blank">Rhino</a> (JavaScript Runtime for the JVM) - its fundamental treatment of asynchronous operations completely differs compared to Node. Simply said, everything in Java is blocking until you make it non-blocking, while in Node (many) operations are non-blocking by default. In Java for example it is a rather common custom to run while-loops until the asynchronous operation sets its condition to false. This repl displays the basic idea: <a href="https://repl.it/@tnobody/non-blocking-vs-blockingjava" target="_blank">V8</a>.
```

Back then, the most common way to handle asynchronous code was via callbacks. They were passed on to the asynchronous operation where they were executed as soon as the initial operation was finished. This approach was useful and acceptable for the usual JavaScript use-case during those days: Handling events in the DOM. But when Node was released and many browser applications became more and more complex, the developers ended up with a so called <a href="http://callbackhell.com/" target="_blank">callback hell</a>. To exit this callback hell, ES6 introduced the <a href="https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise" target="_blank">Promise class</a>. The advantage of a Promise is its ability to chain with other Promises, making it way more elegant to write subsequent, dependent asynchronous operations. A good read on this topic can be found on chapter VIII of <a href="https://exploringjs.com/impatient-js/toc.html" target="_blank">JavaScript for impatient programmers</a>

#### The async wrapper

> You might have noticed that if a testcase in Sakuli v2 is wrapped within `(async () => /*...*/)().then(done)` this construct it is called *Immediately Invoked Function Expression* (IIFE). Basically, it is the definition of a function that is immediately invoked. This pattern is widely used in JavaScript to preserve scopes and namespaces, since every symbol, which is defined at the top level of a script, becomes part of the global namespace and may potentially collide with other scripts. In the era of <a href="https://medium.com/fusebox/beginner-web-developers-use-a-bundler-31ab0c91d2f5" target="_blank">web bundler</a> and <a href="https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/" target="_blank">ES-modules</a> this use-case became more and more irrelevant. IIFE on the other hand still remains important after the introduction of async/await in JavaScript. One caveat of this syntax is, that the `await` keyword is only available within an asynchronous function - because the engine will wrap the whole body of this function within a Promise. That means, that in order to be able to make use of the `await` keyword, your code has to be executed within an asynchronous function. On the top level IIFE, we declare the actual function as `async` and can make use of `await`. Since the result of this function is automatically turned into a Promise, we can invoke `then` and pass the `done` function to it. `done` is a Sakuli specific function, which indicates to Sakuli that the testcase script is finished. There is currently a discussion on the <a href="https://github.com/tc39/proposal-top-level-await" target="_blank">JavaScript proposal in stage 3</a>, which would introduce a *top level await* in JavaScript.

Reviewer: ```suggestion
You might have noticed that if a testcase in Sakuli v2 is wrapped within `(async () => /*...*/)()` this construct it is called *Immediately Invoked Function Expression* (IIFE). Basically, it is the definition of a function that is immediately invoked. This pattern is widely used in JavaScript to preserve scopes and namespaces, since every symbol, which is defined at the top level of a script, becomes part of the global namespace and may potentially collide with other scripts. In the era of <a href="https://medium.com/fusebox/beginner-web-developers-use-a-bundler-31ab0c91d2f5" target="_blank">web bundler</a> and <a href="https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/" target="_blank">ES-modules</a> this use-case became more and more irrelevant. IIFE on the other hand still remains important after the introduction of async/await in JavaScript. One caveat of this syntax is, that the `await` keyword is only available within an asynchronous function - because the engine will wrap the whole body of this function within a Promise. That means, that in order to be able to make use of the `await` keyword, your code has to be executed within an asynchronous function. On the top level IIFE, we declare the actual function as `async` and can make use of `await`.
```

If you want to avoid this construct completely, you can use the `then` function of a Promise. The example code of the <a href="/docs/getting-started/" target="_blank" rel="noopener">Getting started</a> guide would look like this then:

{{< highlight typescript >}}

const testCase = new TestCase();

> _navigateTo("https://sakuli.io")
>     .then(_ => testCase.endOfStep("Open Landing Page", 5, 10))
>     .then(_ => _click(_link("Getting started")))
>     .then(_ => _click(_link("Initialization")))
>     .then(_ => _highlight(_code("npm init")))
>     .then(_ => testCase.endOfStep("Find npm init code sample"))
>     .catch(e => testCase.handleException(e))
>     .then(_ => testCase.saveResult())
>     .then(done);

Reviewer: ```suggestion
_navigateTo("https://sakuli.io")
    .then(_ => testCase.endOfStep("Open Landing Page", 5, 10))
    .then(_ => _click(_link("Getting started")))
    .then(_ => _click(_link("Initialization")))
    .then(_ => _highlight(_code("npm init")))
    .then(_ => testCase.endOfStep("Find npm init code sample"))
    .catch(e => testCase.handleException(e))
    .then(_ => testCase.saveResult());
```
